### ISSUE TITLE
False positive with getEnv double signature

### ISSUE THREAD

**Issue by NicolasCARPi** [vimeo/psalm#9175]
Hello,

## Introduction
`getEnv()` has two signatures:

~~~
'getenv' => ['string|false', 'name'=>'string', 'local_only='=>'bool'],
'getenv\'1' => ['array<string,string>'],
~~~

With an argument, it returns `string|false`, and without an argument it returns an array. Note that it is possible to get an array with `null` and `true|false` as arguments!

## Bug description

There is I believe an issue when you try and cast to string a function call with an argument, warning you that it might be an invalid cast (array -> string).

## How to reproduce

Given the following code:

~~~php
<?php

function buggy(): string
{
    // this will throw a PossiblyInvalidCast error
    // but it contains an argument, and as such won't return an array
    return (string) getEnv('bug');
}
~~~

Running `psalm` on this will show:

INFO: PossiblyInvalidCast - test.php:5:21 - array<string, string> cannot be cast to string (see https://psalm.dev/190)
    return (string) getEnv('bug');

# How can this be fixed

I don't know if it's possible for psalm to specify a different expected signature depending if the first argument is null.

Because if it's not possible, then maybe there is no other choice than to issue a warning.

Please let me know what are your thoughts on this issue.

Best,
~Nico

**Comment by othercorey** [comment]
Seem to happen with `getEnv` but not `getenv`.

**Comment by NicolasCARPi** [comment]
@othercorey You're correct. In fact I don't know why I wrote `getEnv` in the first place, probably the force of habit. The function is all lowercase so there is no reason to do it differently. I'll close this because thare are no issues here.

Best,
~Nico

**Comment by weirdan** [comment]
Functions in PHP are case-insensitive, so `getEnv()`, `getenv()` or even `gEtEnV()` should all result in the same set of issues being reported.
